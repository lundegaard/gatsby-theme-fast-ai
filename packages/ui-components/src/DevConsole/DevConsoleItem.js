import React from 'react';
import PropTypes from 'prop-types';

import Flex from '../Flex';
import Text from '../Text';

const DevConsoleItem = ({ value, label, ...rest }) => (
	<Flex sx={{ py: 2, px: 3, alignItems: 'center' }} {...rest}>
		<Text as="span" fontSize={1} mb={0}>
			{label}
		</Text>
		<Text as="span" ml="auto" mb={0} fontSize={1}>
			{value}
		</Text>
	</Flex>
);

DevConsoleItem.propTypes = {
	label: PropTypes.string,
	value: PropTypes.string,
};

export default DevConsoleItem;
