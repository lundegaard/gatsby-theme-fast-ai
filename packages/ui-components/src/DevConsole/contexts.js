import { createContext } from 'react';

export const DevConsoleApiContext = createContext();
export const DevConsoleLogContext = createContext();
