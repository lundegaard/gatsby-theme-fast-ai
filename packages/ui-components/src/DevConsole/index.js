export { default } from './DevConsole';
export { default as DevConsoleProvider } from './DevConsoleProvider';

export * from './hooks';
