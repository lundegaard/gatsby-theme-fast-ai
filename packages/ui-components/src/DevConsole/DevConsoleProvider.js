import React, { useMemo, useState } from 'react';
import PropTypes from 'prop-types';

import { DevConsoleApiContext, DevConsoleLogContext } from './contexts';

const DevConsoleProvider = ({ children }) => {
	const [log, setLog] = useState([]);

	const api = useMemo(
		() => ({
			log: line => setLog(prevLog => [...prevLog, line]),
			clear: () => setLog(() => []),
		}),
		[]
	);

	return (
		<DevConsoleApiContext.Provider value={api}>
			<DevConsoleLogContext.Provider value={log}>{children}</DevConsoleLogContext.Provider>
		</DevConsoleApiContext.Provider>
	);
};
DevConsoleProvider.propTypes = { children: PropTypes.node };

export default DevConsoleProvider;
