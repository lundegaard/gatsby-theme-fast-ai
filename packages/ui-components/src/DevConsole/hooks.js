import { useContext } from 'react';

import { DevConsoleApiContext, DevConsoleLogContext } from './contexts';

export const useDevConsole = () => useContext(DevConsoleApiContext);
export const useDevConsoleLog = () => useContext(DevConsoleLogContext);
