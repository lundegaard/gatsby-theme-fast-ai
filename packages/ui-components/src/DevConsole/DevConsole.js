import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { keyframes } from '@emotion/core';

import Box from '../Box';
import Flex from '../Flex';
import Text from '../Text';

import DevConsoleItem from './DevConsoleItem';
import { useDevConsoleLog } from './hooks';

const jumpInFromBottom = keyframes`
	0% {
		transform: translateY(200px) scaleY(2.5) scaleX(.2);
		transform-origin: 50% 0%;
		filter: blur(40px);
		opacity: 0;
	}
	100% {
		transform: translateY(0) scaleY(1) scaleX(1);
		transform-origin: 50% 50%;
		filter: blur(0);
		opacity: 1;
	}
}
`;

const easing = 'cubic-bezier(.455, .030, .515, .955)';

/* eslint-disable max-len */
const DropdownArrow = props => (
	<Box
		as="svg"
		width="13px"
		height="9px"
		viewBox="0 0 13 9"
		xmlns="http://www.w3.org/2000/svg"
		{...props}
	>
		<polygon
			fill="currentColor"
			points="10 1.03916875e-13 11.4061556 1.42222591 5.70307778 7.06086213 0 1.42222591 1.40615556 1.03916875e-13 5.70307778 4.24811295"
		/>
	</Box>
);
/* eslint-enable max-len */

const DevConsole = ({ title, initiallyOpened, ...rest }) => {
	const [isOpened, setIsOpened] = useState(initiallyOpened);
	const log = useDevConsoleLog();

	return (
		<Box
			sx={{
				position: 'fixed',
				right: '0',
				bottom: 0,
				animationName: jumpInFromBottom,
				animationDuration: '.5s',
				animationTimingFunction: easing,
				animationFillMode: 'both',
				animationDelay: '.75s',
				backgroundColor: '#3B3B3B',
				color: 'background',
				maxHeight: '100vh',
				minWidth: ['auto', '400px'],
				width: ['100%', '33%'],
			}}
			{...rest}
		>
			<Flex
				sx={{ py: 2, px: 3, alignItems: 'center', cursor: 'pointer' }}
				onClick={() => setIsOpened(!isOpened)}
			>
				<Text as="span" fontFamily="mono" fontSize={1} my={0} sx={{ pointerEvents: 'none' }}>
					{title}
				</Text>
				<DropdownArrow
					sx={{
						marginLeft: 'auto',
						display: 'inline',
						transform: `rotate(${isOpened ? '0' : '180'}deg)`,
					}}
				/>
			</Flex>

			{isOpened && (
				<Box
					sx={{
						borderTopColor: 'background',
						overflowY: 'auto',
						height: '300px',
						borderTopStyle: 'solid',
						borderTopWidth: '1px',
					}}
				>
					{log.map(([label, value], i) => (
						<DevConsoleItem key={i} label={label} value={value} />
					))}
				</Box>
			)}
		</Box>
	);
};

DevConsole.propTypes = {
	children: PropTypes.node,
	initiallyOpened: PropTypes.bool,
	title: PropTypes.node,
};

export default DevConsole;
