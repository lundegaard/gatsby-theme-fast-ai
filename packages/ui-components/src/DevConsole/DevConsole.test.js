import React from 'react';
import { shallow } from 'enzyme';

import DevConsole from './DevConsole';

describe('<DevConsole />', () => {
	const requiredProps = {};

	it('matches snapshot when shallowly rendered', () => {
		const wrapper = shallow(<DevConsole {...requiredProps} />);
		expect(wrapper).toMatchSnapshot();
	});
});
