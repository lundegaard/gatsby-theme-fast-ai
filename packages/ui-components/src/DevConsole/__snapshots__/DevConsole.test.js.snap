// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DevConsole /> matches snapshot when shallowly rendered 1`] = `
<Box
  sx={
    Object {
      "animationDelay": ".75s",
      "animationDuration": ".5s",
      "animationFillMode": "both",
      "animationName": Object {
        "anim": 1,
        "name": "animation-1l35qkc",
        "styles": "@keyframes animation-1l35qkc{
	0% {
		transform: translateY(200px) scaleY(2.5) scaleX(.2);
		transform-origin: 50% 0%;
		filter: blur(40px);
		opacity: 0;
	}
	100% {
		transform: translateY(0) scaleY(1) scaleX(1);
		transform-origin: 50% 50%;
		filter: blur(0);
		opacity: 1;
	}
}
}",
        "toString": [Function],
      },
      "animationTimingFunction": "cubic-bezier(.455, .030, .515, .955)",
      "backgroundColor": "#3B3B3B",
      "bottom": 0,
      "color": "background",
      "maxHeight": "100vh",
      "minWidth": Array [
        "auto",
        "400px",
      ],
      "position": "fixed",
      "right": "0",
      "width": Array [
        "100%",
        "33%",
      ],
    }
  }
>
  <Flex
    onClick={[Function]}
    sx={
      Object {
        "alignItems": "center",
        "cursor": "pointer",
        "px": 3,
        "py": 2,
      }
    }
  >
    <Text
      as="span"
      fontFamily="mono"
      fontSize={1}
      my={0}
      sx={
        Object {
          "pointerEvents": "none",
        }
      }
    />
    <DropdownArrow
      sx={
        Object {
          "display": "inline",
          "marginLeft": "auto",
          "transform": "rotate(180deg)",
        }
      }
    />
  </Flex>
</Box>
`;
